orttraining/orttraining/training_api/module.cc

```diff
@@ -20,6 +24,116 @@ namespace {
+  }
+
+  while (!node_queue.empty()) {
+    auto* node = node_queue.front();
+    node_queue.pop();
+
+    for (auto node_iter = node->InputNodesBegin(); node_iter != node->InputNodesEnd(); ++node_iter) {
+      if (!visited_nodes.count(&(*node_iter))) {
+        node_queue.push(&(*node_iter));
+        visited_nodes.insert(&(*node_iter));
+      }
+    }
+  }
+
+  return visited_nodes;
+}
+
+Status RemoveUnusedNodes(Graph& inference_graph, InlinedVector<const NodeArg*>& output_node_args) {
+  auto reachable_nodes = GetReverseReachableNodes(inference_graph, output_node_args);
+
+  // Get all graph nodes and remove those that are not in the reachable nodes.
+  GraphViewer graph_viewer(inference_graph);
+  for (auto& node : graph_viewer.Nodes()) {
+    if (!reachable_nodes.count(&node)) {
+      inference_graph.RemoveNode(node.Index());
+    }
+  }
+
+  return Status::OK();
+}
+
+Status TransformModelOutputsForInference(std::shared_ptr<Model> inference_model, const InlinedVector<std::string_view>& inference_graph_outputs) {
+  if (inference_graph_outputs.empty()) {
+    // The outputs are not decipherable, so don't transform the graph outputs.
+    return Status::OK();
+  }
+
+  auto& inference_graph = inference_model->MainGraph();
+
+  InlinedVector<const NodeArg*> inference_graph_output_node_args;
+  inference_graph_output_node_args.reserve(inference_graph_outputs.size());
+  for (const auto& output_name : inference_graph_outputs) {
+    const NodeArg* output_node_arg = inference_graph.GetNodeArg(std::string(output_name));
+    ORT_RETURN_IF_NOT(output_node_arg, "Expected graph output for inference graph " + std::string(output_name) +
+                                           " could not be found. Please regenerate the eval graph.");
+    inference_graph_output_node_args.push_back(output_node_arg);
+  }
+
+  // Set the inference graph outputs, and remove any unused nodes.
+  inference_graph.SetOutputs(inference_graph_output_node_args);
+  ORT_RETURN_IF_ERROR(RemoveUnusedNodes(inference_graph, inference_graph_output_node_args));
+
+  ORT_RETURN_IF_ERROR(inference_graph.Resolve());
+
+  return Status::OK();
+}
+
+InlinedVector<std::string_view> ParseInferenceGraphOutputsFromMetadata(const ONNX_NAMESPACE::ModelProto& model_proto) {
+  std::string inference_graph_outputs_string;
+  for (const auto& model_metadata : model_proto.metadata_props()) {
+    if (model_metadata.key() == Module::InferenceGraphOutputsMetadataString) {
+      inference_graph_outputs_string = model_metadata.value();
+      break;
+    }
+  }
+
+  return onnxruntime::utils::SplitString(inference_graph_outputs_string, ",");
+}
+
+Status TransformModelInputsForInference(std::shared_ptr<Model> inference_model, const std::unordered_map<std::string, std::shared_ptr<Parameter>>& named_parameters, const DataTransferManager& data_transfer_manager) {
+  std::vector<const NodeArg*> user_graph_inputs;
+  for (auto& graph_input_node_arg : inference_model->MainGraph().GetInputs()) {
```

Comment:
nit: get inference_model->MainGraph() once, and use it multiple times as TransformModelOutputsForInference did.